iXBRLViewerPlugin/viewer/src/js/outline.js

```diff
@@ -0,0 +1,153 @@
+// Copyright 2021 Workiva Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import $ from 'jquery'
+
+// DocumentOutline chooses a fact for each presentation group (ELR) that
+// represents the start of that ELR.  This is done by deciding which ELRs each
+// fact participates in (see factInGroup()) and then finding the longest
+// continuous run of facts in document order that participate in each ELR.
+export function DocumentOutline(report) {
+    this._report = report;
+    var facts = report.facts().sort((a, b) => a.ixNode.docOrderindex - b.ixNode.docOrderindex);
+    var runLength = {};
+    var runStart = {};
+    var longestRun = {};
+    var longestRunStart = {};
+    this.dimensionMap = this._buildDimensionMap();
+    const elrs = report.relationshipGroups("pres");
+    for (const f of facts) {
+        if (f.isHidden()) {
+            continue;
+        }
+        for (const elr of elrs) {
```

Comment:
I wonder how this fact/elr/dimension looping performs for reports with tends of thousands of facts.  @derekgengenbacher-wf Can you be sure to test this against large doc?

Reply:
I tested this and it definitely slows the loading of the js on the order of about 10 seconds.

Reply:
How much slower is that?  Do we need to change it to only do it lazily or find some other way to speed it up?

Reply:
Having run it a few more times in both wk-dev and locally, the time is negligible. The first few times I ran I had an outlier that took a few seconds longer than the rest of the attempts. I was also running a validation in the back ground during that run.